### ISSUE TITLE
Streaming response handler

### ISSUE THREAD

**Issue by piercefreeman** [piercefreeman/gpt-json#9]
Motivated by https://github.com/piercefreeman/gpt-json/pull/8, here are some general notes on introducing a streaming API to gpt-json:

## API

The streaming API is basically a greedy return function; it can be used in cases where users need to display pending results to users (like ChatGPT's typing animation) or when the client user needs to process partial objects as soon as they are reasonably available. Either a callback or an async iterator are the most natural paradigms for this interaction. We propose using the async iterator because it allows control flow to be kept in one place without the need to switch to a wholly other function and maintain shared state in some global store.

Example caller:

```python
class SentimentSchema(BaseModel):
    thought: str
    sentiment: str
    justification: str | None = None

gpt_json = GPTJSON[SentimentSchema](API_KEY)
messages = [
            GPTMessage(
                role=GPTMessageRole.SYSTEM,
                content=SYSTEM_PROMPT,
            ),
        ]

async for partial_result in gpt_json.stream(messages=messages):
    print(partial_result)
```

## Validation

These `gpt_json.stream` calls should have the following properties:

1. Yield `partial_results`, which are instances of SentimentSchema. We recognize that SentimentSchema might have some keys that are required, and instance creation will fail until these are populated. In this case, `thought` and `sentiment` are both required. `gpt_json.stream` should therefore only start streaming results once we have both `thought` filled and an initial value for `sentiment`. The sentiment value can be partially filled (ie. one word out of 10). But both must have some non-null value as specified in the BaseModel.
2. Yield a new `partial_result` every time we get a new token back from the server. We can optionally add some smart-queuing logic so if multiple tokens come back at the same time we will only return one object that incorporates all the changes. There's no need to process token-by-token if more are already available.
3. Optionally yield the `updated_key` that was most recently updated. This will allow client callers to take conditional logic based on the current key that is being processed.

3.a. Either yield another value in the tuple like this: 

```python
async for partial_result, updated_key in gpt_json.stream(messages=messages, yield_keys=True):
  print(partial_result)
```

3.b. Instead of yielding `partial_results` as SentimentSchema, we yield them as PartialObject, which is a wrapper schema.

```python
class PartialObject(Generic[T]):
  updated_key: str
  partial_obj: T
```

4. Ensure the behavior of fetching and streaming is properly async and can yield control flow back to the program as we are waiting for results, using something like `httpx.stream`. I believe ChatCompletion.acreate is available with the `stream=True` endpoint and implements its logic like this but we'll need to double check.

**Comment by craterink** [comment]
looks good, thanks for writing this up

it seems like the 2 main usecases of streaming are:
1) visual rendering (better UX than waiting for a long JSON object to complete)
2) message-passing partial results to another async module for better/fancier distributed processing

does that sound right?

also, a couple thoughts on the design:
A) smart-queueing logic sounds maybe overdose for v0, but agree it's a neat idea for some usecases of streaming
B) I like the example caller & 3a designs but I think we'd also want an easy way to know we've reached a completed value. i.e. we might want to provide a signal that the value is finished being streamed. something like the below

```
async for partial_result, event, event_data in gpt_json.stream(messages=messages, yield_events=True):
  print(partial_result, event, event_data)
```

which prints

```
result, "value_updated",  {"key": "key whose value was updated", "update": "content delta"}
result, "value_updated",  {"key": "key whose value was updated", "update": "content delta"}
result, "value_completed", {"key": "key whose value was completed", "value": "full completed value"}
```

this gives you flexibility to ignore `value_updated` events and wait for completed values if you'd like

**Comment by piercefreeman** [comment]
@craterink 

Curious to hear more what you're thinking around use case (2), maybe with a motivating example. I was originally proposing something more like (3) if a client needs a partial payload as soon as possible, like to upsert data into a SQL table or kick off some meta model based on the results, they can. Mostly serves to decrease latency if you don't need the whole payload to do something useful.

A). Agreed, out of scope for a V0. So the actual design should be yielding after every new token result is retrieved from the server?
B). I think the client is able to retrieve the `value` here directly from the result payload. But I do agree some lifecycle notifications might be helpful. Given this complexity I'm leaning towards 3.b. myself:

```python
class StreamEventEnum:
    KEY_UPDATED
    KEY_COMPLETED

class StreamingObject(Generic[T]):
  updated_key: str
  event: StreamEventEnum
  partial_obj: T
```

I believe that's inclusive of both cases you've mentioned.

**Comment by craterink** [comment]
re (3), that is a much clearer explanation of what I was going for with (2) lol. so we're good on that part I think.

A) Sounds good!

B) hmm this works. but, if you only want diff(prev_value, curr_value) -- which is what I needed for my CLI usecase -- you'd have to maintain the previous partial_obj state and take the diff and that feels a little clunky. why not include the update in the object? low cost right

**Comment by piercefreeman** [comment]
B. Does the API ever delete / modify previous values in the sequence, like in the resolution of beam search? I recall observing some behavior like this in the UI but the stream documentation has no indication that sequences won't be merely additive.

In this case the content delta seems relevant (but IMO I'd only do it when tied to a StreamEventEnum.KEY_UPDATED). KEY_COMPLETED should just be an indication that the JSON payload has officially been closed and no more values will be appended for that key - so no delta is relevant.

**Comment by craterink** [comment]
closed with #8
